MapleServer2/Tools/InventoryController.cs

```diff
@@ -1,46 +1,153 @@
 ﻿using System;
 using MapleServer2.Types;
 using MapleServer2.Packets;
-using System.Diagnostics;
+using MaplePacketLib2.Tools;
 using MapleServer2.Servers.Game;
+using Maple2Storage.Types;
 
 public class InventoryController
 {
-	public static void Add(GameSession session, Item item)
-	{
-        // Checks if item is stackable or not
+    // Adds item
+    public static void Add(GameSession session, Item item, Boolean isNew)
+    {
+        // Checks if item is stackable or not.
         if (item.SlotMax > 1)
         {
             foreach (Item i in session.Player.Inventory.Items.Values)
             {
-                // Continue if inventory item id doesn't match or it is at max amount
+                // Checks to see if item exists in database (dictionary)
                 if (i.Id != item.Id || i.Amount >= i.SlotMax)
                 {
                     continue;
                 }
-                // Updates inventory for item amount overflow
+                // Updates inventory for item amount overflow.
                 if ((i.Amount + item.Amount) > i.SlotMax)
                 {
-                    int added = i.SlotMax - i.Amount; // For marking item new with correct added amount
-
                     item.Amount = item.Amount - (i.SlotMax - i.Amount);
                     i.Amount = i.SlotMax;
                     session.Send(ItemInventoryPacket.Update(i.Uid, i.Amount));
-                    session.Send(ItemInventoryPacket.MarkItemNew(i, added));
                 }
                 // Updates item amount
                 else
                 {
                     i.Amount = i.Amount + item.Amount;
                     session.Send(ItemInventoryPacket.Update(i.Uid, i.Amount));
-                    session.Send(ItemInventoryPacket.MarkItemNew(i, item.Amount));
                     return;
                 }
             }
         }
-        // Add item to inventory if cannot stack any further
-        session.Player.Inventory.Add(item); 
-        session.Send(ItemInventoryPacket.Add(item));
-        session.Send(ItemInventoryPacket.MarkItemNew(item, item.Amount));
+        session.Player.Inventory.Add(item); // adds item into internal database
+        session.Send(ItemInventoryPacket.Add(item)); // sends packet to add item clientside.
+
+        if (isNew)
+        {
+            session.Send(ItemInventoryPacket.MarkItemNew(item, item.Amount)); // Marks Item as New
+        }
```

Comment:
This change to how MarkItemNew is handled is wrong and also adds a needless parameter to the function, please revert. With this implementation, stacks won't be marked as new.